File: lib/python/opentoken/src/main/opentoken/attributes/validation/date_range_validator.py
"""
Copyright (c) Truveta. All rights reserved.
"""

from datetime import datetime, date
from typing import Optional
from opentoken.attributes.validation.serializable_attribute_validator import SerializableAttributeValidator


class DateRangeValidator(SerializableAttributeValidator):
    """
    A generic validator that asserts that a date value is within an acceptable range.

    This validator checks that dates are:
    - Not before the specified minimum date (if provided)
    - Not after the specified maximum date (if provided)
    - If use_current_as_max is True, uses today's date as the maximum

    If the date is outside the specified range, the validation fails.
    """

    # Supported date formats for parsing
    POSSIBLE_INPUT_FORMATS = [
        "%Y-%m-%d",
        "%Y/%m/%d",
        "%m/%d/%Y",
        "%m-%d-%Y",
        "%d.%m.%Y"
    ]

    def __init__(self, min_date: Optional[date] = None, 
                 max_date: Optional[date] = None,
                 use_current_as_max: bool = False):
        """
        Initialize the validator with a date range.

        Args:
            min_date: The minimum allowed date (inclusive), or None for no minimum
            max_date: The maximum allowed date (inclusive), or None for no maximum
            use_current_as_max: If True, uses today's date as the maximum date
        """
        self._min_date = min_date
        self._use_current_as_max = use_current_as_max
        self._max_date = max_date

    @property
    def min_date(self) -> Optional[date]:
        """Get the minimum allowed date."""
        return self._min_date

    @property
    def max_date(self) -> Optional[date]:
        """Get the maximum allowed date."""
        if self._use_current_as_max:
            return date.today()
        return self._max_date

    @property
    def use_current_as_max(self) -> bool:
        """Check if using current date as maximum."""
        return self._use_current_as_max

    def eval(self, value: str) -> bool:
        """
        Validate that the date value is within acceptable range.

        Args:
            value: The date string to validate

        Returns:
            True if the date is within acceptable range, False otherwise
        """
        if not value or not value.strip():
            return False

        parsed_date = self._parse_date(value.strip())
        if parsed_date is None:
            return False

        # Check minimum date if specified
        if self._min_date is not None and parsed_date < self._min_date:
            return False

        # Check maximum date
        max_date_to_check = self.max_date
        if max_date_to_check is not None and parsed_date > max_date_to_check:
            return False

        return True

    def _parse_date(self, value: str) -> Optional[date]:
        """
        Parse the date string using various supported formats.

        Args:
            value: The date string to parse

        Returns:
            Parsed date object, or None if parsing fails
        """
        for date_format in self.POSSIBLE_INPUT_FORMATS:
            try:
                parsed_datetime = datetime.strptime(value, date_format)
                return parsed_datetime.date()
            except ValueError:
                continue
        return None

File: lib/java/opentoken/src/main/java/com/truveta/opentoken/attributes/validation/DateRangeValidator.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.attributes.validation;

import java.text.ParseException;
import java.time.LocalDate;
import java.time.ZoneId;
import java.util.Date;
import java.util.Locale;

import org.apache.commons.lang3.time.DateUtils;

import lombok.Getter;
import lombok.Setter;

/**
 * A generic validator that asserts that a date value is within
 * a configurable date range.
 * 
 * This validator can be configured with:
 * - Minimum date (inclusive) - can be null for no lower bound
 * - Maximum date (inclusive) - can be null for no upper bound
 * - Use current date as maximum - dynamically sets max to today
 * 
 * If both minDate and maxDate are null and useCurrentDateAsMax is false,
 * the validator will only check that the date can be parsed.
 */
@Getter
@Setter
public final class DateRangeValidator implements SerializableAttributeValidator {

    private static final long serialVersionUID = 1L;

    // Supported date formats for parsing
    private static final String[] POSSIBLE_INPUT_FORMATS = new String[] {
            "yyyy-MM-dd", "yyyy/MM/dd", "MM/dd/yyyy",
            "MM-dd-yyyy", "dd.MM.yyyy"
    };

    private final LocalDate minDate;
    private final LocalDate maxDate;
    private final boolean useCurrentDateAsMax;

    /**
     * Creates a DateRangeValidator with specified minimum and maximum dates.
     * 
     * @param minDate the minimum allowed date (inclusive), or null for no lower bound
     * @param maxDate the maximum allowed date (inclusive), or null for no upper bound
     */
    public DateRangeValidator(LocalDate minDate, LocalDate maxDate) {
        this.minDate = minDate;
        this.maxDate = maxDate;
        this.useCurrentDateAsMax = false;
    }

    /**
     * Creates a DateRangeValidator with specified minimum date and option to use current date as max.
     * 
     * @param minDate the minimum allowed date (inclusive), or null for no lower bound
     * @param useCurrentDateAsMax if true, uses current date as maximum; otherwise no upper bound
     */
    public DateRangeValidator(LocalDate minDate, boolean useCurrentDateAsMax) {
        this.minDate = minDate;
        this.maxDate = null;
        this.useCurrentDateAsMax = useCurrentDateAsMax;
    }

    /**
     * Creates a DateRangeValidator with no bounds (will only validate parseable dates).
     */
    public DateRangeValidator() {
        this.minDate = null;
        this.maxDate = null;
        this.useCurrentDateAsMax = false;
    }

    /**
     * Validates that the date value is within the configured range.
     * 
     * @param value the date string to validate
     * @return true if the date is within the configured range, false otherwise
     */
    @Override
    public boolean eval(String value) {
        if (value == null || value.trim().isEmpty()) {
            return false;
        }

        try {
            // Parse the date using various supported formats
            Date date = DateUtils.parseDateStrictly(value, Locale.ENGLISH, POSSIBLE_INPUT_FORMATS);

            // Convert Date to LocalDate
            LocalDate localDate = date.toInstant().atZone(ZoneId.systemDefault()).toLocalDate();

            // Check minimum date
            if (minDate != null && localDate.isBefore(minDate)) {
                return false;
            }

            // Check maximum date
            LocalDate effectiveMaxDate = useCurrentDateAsMax ? LocalDate.now() : maxDate;
            return effectiveMaxDate == null || !localDate.isAfter(effectiveMaxDate);

        } catch (ParseException e) {
            // If the date cannot be parsed, it's invalid
            return false;
        }
    }
}
